### ISSUE TITLE
Watch a directory recursively

### ISSUE THREAD

**Issue by roma-glushko** [ProCern/asyncinotify#2]
Hey folks 👋 

Is it possible to watch the directory recursively? So when I do `add_watch("/home/roma/")`, I receive events from:
- `/home/roma/*` files per se
- `/home/roma/docs` and any other existing subdirectory of the path
- `/home/roma/justaddeddir/*` - any new directory created after we have started watching the root dir

I assume that this is a common case for inotify usage and would be great to have in `asyncinotify`.

Thank you in advance!

**Comment by Taywee** [comment]
Not integrated into the actual library, but [the sole example in the examples section is an example of this functionality](https://github.com/absperf/asyncinotify/blob/c5590c5c472c3b1ee2c5a408ffba04889c6b65d4/examples/recursivewatch.py).

[This has been asked for before, in the GitLab mirror](https://gitlab.com/Taywee/asyncinotify/-/issues/4), and [more than once, at that](https://gitlab.com/Taywee/asyncinotify/-/issues/15).

I'm not against the functionality, but I'm hesitant to add it into the module itself, because:

* The job of this library is mostly to just bring a Pythonic asyncio interface to inotify, and this functionality isn't natively part of inotify, so it would be an extra nicety.
* [There are subtle behavioral things that can throw this off](https://github.com/absperf/asyncinotify/blob/c5590c5c472c3b1ee2c5a408ffba04889c6b65d4/examples/recursivewatch.py#L38-L59).
* There are open questions with how it interacts with existing `Inotify` objects, if at all.
  * Should this be behavior directly on `Inotify`?
    * I think it shouldn't be.  `Inotify` is supposed to be a direct representation of the underlying interface, with just enough added to idiomatically work with it as an asynchronous Python interface.
  * Should the interface completely wrap an `Inotify` instance like the example?
    * This works, unless you want to watch a directory recursively at the same time as a non-recursive one with the same `Inotify` instance.
  * Should the interface just take in another `Inotify` instance?
    * I'd probably prefer this, then you could add things to `Inotify` as you like and then just iterate the `RecursiveDirectoryWatcher` and get all the recursive and non-recursive events as you like them.

You're right that it's a common case for inotify, and it sounds like a simple add, but a lot of little things need to be accounted for, even before considering race conditions and timing-sensitive things.  For me to integrate the functionality, it needs to be the kind of thing that anybody who wants a recursive inotify watcher could just reach for and would never have to work around or reimplement because it does things slightly different than they would accept.

I'm going to keep this issue open, as the GitLab one, because I do think that the functionality would fit well in here, and I'm more than willing to review PRs for it.

**Comment by zhuowang10** [comment]
Hi Taywee,
I improved your sample in my project, and covered more use cases. The solution turns out is not complicated. I think I can contribute that to this project. Check in here to see if you're interested.
https://github.com/zhuowang10/syncservers/blob/main/syncservers/recursive_watcher.py

**Comment by Taywee** [comment]
@zhuowang10 Yes, that largely does look like an improvement.  I'm a little embarrassed for using `is_dir` instead of `Mask.ISDIR` in my version.

My only concern is that directory handling is racy.  From the inotify man page, if we rename a directory, destroying and recreating the inotify can cause events to be lost in the interim, and pending events can then be correlated to non-existing watch descriptors.  `MOVED_FROM` and `MOVED_TO` have a cookie to link them, but they can have other events between them, and there can be a delay before the latter (or it can even be absent, if moved out of the watched directory).  A timer can be used to track this, but then you can get events for a directory that shouldn't be watched anymore.

That said, I consider your version a definite improvement, and would be happy to merge your changes into this codebase if you open a PR with the modifications, particularly if the race conditions are mentioned in a comment (probably replacing [these lines](https://github.com/ProCern/asyncinotify/blob/c5590c5c472c3b1ee2c5a408ffba04889c6b65d4/examples/recursivewatch.py#L47-L53)).

**Comment by zhuowang10** [comment]
yeah, you're right about the racing conditions: events won't be captured right after a new folder was added, or still sending events right after a folder was moved out of monitored folders. I think this is how inotify works, what we can do, as you suggested, is to make the expectations clear.

per is_dir, it is actually needed, caused by the racing, only one folder creation event is captured when cascading folders were created, so is_dir is needed to verify those un-captured folders.

**Comment by Taywee** [comment]
Yes, for the `get_directories_recursive`, but I meant the use of `ISDIR` in [this line](https://github.com/zhuowang10/syncservers/blob/12a9753f64bf7a11e7a137ea7566808fd6c9317d/syncservers/recursive_watcher.py#L34).  The `path.is_dir` performs a redundant stat in the original, when the information is already in the event mask.

**Comment by zhuowang10** [comment]
PR is in. :)

**Comment by TaylorRichberger** [comment]
@zhuowang10 
And it's merged.  Thank you.  I like that it's a separate class and not worked into the main classes.  I'm happy with that compromise.  I would never want it to be integrated into the core classes (which should always remain nothing more than inotify translated into clean, idiomatic, async Python), but it's a nice extra to have on hand.

**Comment by zhuowang10** [comment]
I'm glad I can contribute while leveraging it for my own projects.
Can you release a new version to pypi to include this change?

**Comment by TaylorRichberger** [comment]
@zhuowang10
Sure.  I've just released it now as v4.1.0

**Comment by zhuowang10** [comment]
Hi Taylor,
Although the tag 4.1.0 is up-to-date, the pypi 4.1.0 install seems not containing recursive watcher. This is what I tried:

```bash
python3 -m pip install asyncinotify==4.1.0
Collecting asyncinotify==4.1.0
  Using cached asyncinotify-4.1.0-py3-none-any.whl (16 kB)
Installing collected packages: asyncinotify
Successfully installed asyncinotify-4.1.0

wc -l < asyncinotify/__init__.py 
651
```

**Comment by TaylorRichberger** [comment]
@zhuowang10 Good catch.  I honestly have no idea what happened there.  I bumped the version after merging your pull request, so it's weird that that version ended up without the watcher.  This has been corrected now via v4.2.0, which I've double checked and definitely contains the latest version.
